Implemented the trans and blocktrans tags

ehoogerbeets opened the pull request:
Notes:
- Implemented unit tests for both tags which show how they work in normal situations, and how they throw exceptions in the right situations
- Added Loader.findResourcesDir() so that callers can find the first resources directory along the same path where the loaders load the templates. This way, the translations and the templates can go together, where-ever they are located.
- Modified the environment to call findResourcesDir once.
- This code does not depend on ilib for translation, though of course I think it should be the recommended solution! ;-)

ehoogerbeets committed:
Implemented the trans and blocktrans tags

Notes:

- Implemented unit tests for both tags which show how they work
in normal situations, and how they throw exceptions in the right
situations

- Added Loader.findResourcesDir() so that callers can find the
first resources directory along the same path where the loaders
load the templates. This way, the translations and the templates
can go together, where-ever they are located.

- Modified the environment to call findResourcesDir once.

- This code depends on ilib for translation. Debating on whether
or not this should be the case. On the one hand, more i18n should
happen in the filters, as the support is lacking there. Ilib can
help with that. On the other, it would be cleaner if reinhardt
could stand alone...

ehoogerbeets commented:
This is the first commit. Not ready to pull it into your repo yet. Still need to update the documentation and decide about the ilib issue. (Should reinhardt depend on it or not?)

oberhamsi commented:
thanks for putting the code online. my thoughs:
- if it depends on ilib it should be a package.json dependency (i think you plan on putting ilib on packages.ringojs.org anyhow)
- put the new Tags into tags/i18n.js (not tags/misc.js) and add the new module here: https://github.com/orfon/reinhardt/blob/master/lib/parser.js#L17
- three spaces indentation please (yes, that's odd. reinhardt needs a contrib document)

oberhamsi commented on test/trans.js:
```diff
@@ -0,0 +1,174 @@
+		var x = env.getTemplate('page12.html').render(c);
+		assert.equal(true, false);
+	} catch (e) {
+		// success because it is supposed to throw an exception
```
looks like you want `assert.throws()`  here http://ringojs.org/api/master/assert/index.html#throws

ehoogerbeets committed:
Incorporate review comments from Simon

- use 3 spaces for indent
- use assert.throws for the tests
- move trans and blocktrans tag implementations to their own i18n.js
file

ehoogerbeets committed:
Fix a few more indentation problems.

ehoogerbeets committed:
More review comment fixes

ehoogerbeets commented on test/trans.js:
```diff
@@ -0,0 +1,129 @@
+var {Environment} = require('../lib/environment');
+var {Template} = require("../lib/template");
+var {TemplateSyntaxError} = require('../lib/errors');
+var ilib = require("../../ilib");
```
This path works for tests and in ringo on a web server, but only when ilib is checked out right next to reinhardt. In ringo running on jetty, you can use simply require("ilib") and it will find it properly in the packages dir, but that doesn't work for the tests. Is there some way to add to the path where the test framework looks for packages?

oberhamsi replied:
> Is there some way to add to the path where the test framework looks for packages?

yeah, there's a command line arg:

```
ringo --modules /path/to/packages/
```

or you can install the package globally with `rp install --global`

ehoogerbeets committed:
Now does not depend on ilib for translation

If you create a Resources class with a translate method, the code
will use that to do translation. Also, you can pass a resBundleFactory
method to the environment as you construct it to create new bundles
as necessary. Of course, all of it can be implemented using ilib,
but it is up to the app to choose which translation solution they
prefer as long as they wrap it in the interface mentioned above.

ehoogerbeets committed:
Updated the documentation for the trans/blocktrans tags

Updated the documentation to talk about the new trans and blocktrans
tags, how to use them, and a few pitfalls to avoid.

ehoogerbeets commented:
Hi @oberhamsi, I think this PR is ready for review now. Please take a look. Please read the updated doc careful to make sure I got it all correct. I'm going to put a few questions in line comments there for you.

If it is okay, then I will work with it for a bit from my own master branch while developing my own site. I'll probably find a few bugs or upgrades here and there which I can fix. Once I'm comfortable that it is all stable, we can merge my master branch back to the main repo again.

ehoogerbeets committed:
Fixed whitespace issues.

tab -> 4 spaces, trailing whitespace eliminated.

ehoogerbeets commented on docs/templates.md:
```diff
@@ -601,3 +611,336 @@ access to the comments template tags and filters. The child template is
+the built-in filters. You might look for Ringo packages on RP that implement
+locale-sensitive filter plugins or implement them yourself.
+
+Just so that you are aware when you use them, the following filters perform
```
Do we want to list these i18n problems here? This is part of the big list of i18n problems I talked about previously.

oberhamsi replied:
Yes, I think we could even call it more explicitly: "Problems with i18n in Reinhard" or so :) Not sure the detailed list of problematic filters makes sense. I created https://github.com/orfon/reinhardt/issues/17 and https://github.com/orfon/reinhardt/issues/16

And IMO this is still so much info that it needs its own i18n.md. Templates.md should just give a short overview - maybe the first paragraph - and then link to i18n.md. Like the other sections above it.

ehoogerbeets commented on docs/templates.md:
```diff
@@ -601,3 +611,336 @@ access to the comments template tags and filters. The child template is
+Because you have to put the locale setting into the context in order to do translation,
+this setting is also available to use in conditional statements. Here is
+an example:
+
```
Is this the right syntax for the if conditions?

oberhamsi replied:
yes. `==` would work too.

ehoogerbeets commented on docs/templates.md:
```diff
@@ -601,3 +611,336 @@ access to the comments template tags and filters. The child template is
+in the render context:
+
+    app.get('/:page', function(request, page) {
+        return env.renderResponse(page, {
```
Currently, renderResponse calls the resource bundle factory to get the resource bundle and adds it to the context. If you just call node.render directly, this does not happen. Is this okay?

The problem is that the render() methods only take a context argument, so there is no way to pass it per-request information except via the context. Something has to wrap the render calls and put the resource bundles into the context before calling render. I'm not sure that it is thread-safe though. Is each request a separate thread? Is the context specific to a single thread so we don't have to worry about race conditions?

oberhamsi replied:
> The problem is that the render() methods only take a context argument, so there is no way to pass it per-request information except via the context.

I would do just that: pass the locale as context.locale. The Environment doesn't have to be involved in i18n (execpt for holding i18n configuration): it just passes the context down to the Tags as it did before. And the Tags can lookup the resource bundle themselves. I admit it's not obvious that Tags can access the environment with `this.env`. But because of this, the `if` from Environment.renderResponse ( https://github.com/ehoogerbeets/reinhardt/blob/transtag/lib/environment.js#L50 ) can move into the render methods of BlockTransTag and TransTag.

If we do it like that, everything about i18n is contained in the i18n Tags. And passing locale via the context works in all cases: Env.render, Template.render and the convinience method Env.renderResponse. It makes `context.locale` a bit special but i can live with that.

Unrelated to the above:

> I'm not sure that it is thread-safe though. Is each request a separate thread? Is the context specific to a single thread so we don't have to worry about race conditions?

Yes, each request is in its own thread and the context is not shared by threads. The docs warn about storing per-request information on Tag instances, if you are worried about that.  i18n tags don't do that so they are fine.

oberhamsi commented on lib/environment.js:
```diff
@@ -21,6 +21,9 @@ exports.Environment = Environment = function(options) {
       filters: options.filters,
       tags: options.tags
    };
+   if (options.i18n) {
```
If you remove this `if` and its body, then the i18n configuration is still available at `env.config.i18n`. Tags should put their configuration there and not directly at the environment.

oberhamsi commented on lib/environment.js:
```diff
@@ -32,6 +35,9 @@ exports.Environment = Environment = function(options) {
    delete options.debug;
    this.config = options;
+   
+   this.resourcesDir = this.loader && typeof(this.loader.findResourcesDir) === "function" ? this.loader.findResourcesDir() : "./resources";
```
I don't understand this: The filesystem Loader looks for a given file in multiple base paths and returns the first match. It's useful for overriding base templates with more specialized variations. ("mysite/base", "mysite/customer1", ...) I don't know if this is helpful for i18n resources.

I wonder if we can drop env.resourcesDir and Loader.findResourcesDir. The resource dir can be looked up when creating the i18n configuration for the Environment, right?

ehoogerbeets replied:
Yeah, I agree, it's not quite right yet. Maybe you help me find a better way? Here's what I'm trying to accomplish. The idea is that each set of templates should come with its own resources. This should be true of mounted applications as well as templates loaded via a different loader. 

Advantages: 
- you can easily add or subtract sets of templates or mounted apps. Each set comes with everything it needs for translation. 
- sets and mounted app can easily be shared
- any loc tool that extracts strings for a set of templates only needs to scan that set of templates, not anything else.

Disadvantages:
- strings are distributed, making them harder to find when debugging
- might have some duplication
- the added complexity makes it a little harder to understand and use

Based on this, I was trying to find, for any particular template, where it was loaded from so that I could then go load its resources from the root of the same location.

oberhamsi replied:
I looked at how django and rails do translation resource lookup, and I think we don't have to provide any new functionality (findResourceDir et al.) to keep the advantages you mention.

This is how I imagine it:

In a multi-app scenario, each app would configure its own template environment with its own resource path (a relative path within the app's package). The translation resources would be distributed - per application - but not shared. A super-app could not override sub-app translations but on the other hand the sub-apps wouldn't even have to agree on one translation package (e.g. ilib and gettext).

This would still leave the disadvantage of dealing with distributed translations. Django has rosetta [1] to make that easier. I think this problem is out-of-scope for our first implementation.

[1] http://django-rosetta.readthedocs.org/en/latest/usage.html#generate-a-batch-of-files-to-translate

somewhat related: A location tool extracting the translatable strings would also operate per application: given a list of file extensions and paths, the extractor runs the files through a new subclass of parser, which only cares about the translation nodes and will output the skeleton translation resources.

oberhamsi commented on docs/templates.md:
```diff
@@ -601,3 +611,336 @@ access to the comments template tags and filters. The child template is
+The parameter of the ``trans`` tag should be a quoted source string written in the
+source language of the template.
+
+N.B. The templates should be encoded in UTF-8 for Reinhardt to work properly.
```
i created https://github.com/orfon/reinhardt/issues/18. no one has complained so far :)

oberhamsi commented:
You could list `trans` and `blocktrans` in docs/tags.md. Maybe just links to i18n.md would already make them easier to discover.

ehoogerbeets committed:
Changes from review comments from Simon

Changes based on review comments from Simon.
* Moved the resource bundle factory call into the trans and blocktrans nodes
* Moved the globalization docs into i18n.md and linked it to the other docs
* Updated tags.md to add notes about trans and blocktrans

ehoogerbeets commented:
@oberhamsi final review?

oberhamsi commented:
:+1: looks great. seems obvious to say it: but i especially like the docs ;)

please open PR on orfon/reinhardt if you want the github karma points - or i can merge from here.
